Sources/SwiftSyntax/Assert.swift

```diff
@@ -0,0 +1,51 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// How to choose `assert` vs. `precondition`:
+///  - Wherever possible, it is preferrable to emit a diagnostic instead of
+///    using `precondition`. This way the parser won't crash if the condition is
+///    violated.
+///  - If you think the diagnostic added above should never be emitted, it is
+///    fine to also emit an `assertionFailure` in debug builds to make it easier
+///    to debug the unexpected diagnostic.
+///  - If in doubt always use `precondition`
+///  - `assert` should only be used if checking the assertion has a non-trivial
+///    cost and provides little benefit in terms of safety in release builds.
+
+// MARK: - Assert
+
+/// An assertion that is active in DEBUG builds, just like `Swift.assert` and
+/// additionally if assertions are explicitly requested by setting the
+/// `SWIFTSYNTAX_ENABLE_ASSERTIONS` conditional compilation flag.
+/// Use this instead of `precondition` in places where the assertion has a
+/// non-trivial cost but provides little value in release builds.
+@_transparent
+public func assert(_ condition: @autoclosure () -> Bool, _ message: @autoclosure () -> String = String(), file: StaticString = #file, line: UInt = #line) {
+  #if SWIFTSYNTAX_ENABLE_ASSERTIONS
+  precondition(condition(), message(), file: file, line: line)
```

Comment:
You could change this to fatalError rather than precondition so we get messages still. Or not, probably doesn't really matter.